### ISSUE TITLE
Generating class which name is 'Option'

### ISSUE THREAD

**Issue by BBartosz** [twilio/guardrail#438]
I have swagger docs which has some conflicting part:
```json
"option": {
          "description": "Some options blabla",
          "type": "object",
          "properties": {
            "inventorySource": {bla bla}
           }
}
```

So by default generated case class for this object looks like this:
`case class Option(...)`
which obviously conflicts with scala Option. Is it a way to generate definiton class with some prefix to avoid collision like this ?

**Comment by ChengGit** [comment]
hi @BBartosz, in my humble opinion, code-gen should be ignostice for any class defined inside user domain/package, just like compiler allowing user to define `my.package.my.module.Option`. In terms of naming-conflictions, may I suggest to either giving a more descriptive class name, or rename it as part of import, i.e. `import myPackage.definitions.myRequest.{Option => GeneratedOption}`
May I have your opinions on this?

**Comment by BBartosz** [comment]
Maybe I didnt specify the problem enough. So its not me who specify name of class thats the problem. Guardrail generates it based on name of field, so if the field is called unfortunately for me like scala reserved keywords we will have problems like this everytime. so given swagger documentation:
```
"option": {
          "description": "Some options blabla",
          "type": "object",
          "properties": {
            "inventorySource": {bla bla}
           }
}
```
generated class will look like:

```
case class Option(someField, someField)
```

Which results in conflict. And I cant change whole api contract of some api just because in some language there is naming conflict, it doesnt sound right, do you agree? So I dont understand a bit your solution with changing imports. Yeah I could even rename `case class Option` to be `case class _Option` but then should i rename it every time i recompile the project? Or I am missing something? I think generating tool should give me some flexibility how i want generated code to look like.

**Comment by blast-hardcheese** [comment]
@BBartosz This may actually already be solved by #434! If you wouldn't mind running your spec via the CLI in that branch, I'd appreciate it. My goal is to review and merge that sometime this weekend.

**Comment by tomasherman** [comment]
im not sure those changes are gonna help, i think there are still some cases where FQNs are not used - but i def agree that it would be a good idea to do so and changes in #434 have most likely provided the tools to fix it in other places aswell

**Comment by BBartosz** [comment]
@tomasherman  @blast-hardcheese If those changes from #434 add package to class name that would solve my issue. How to publish guardrail locally so i can run it against my spec? I guess through `sh cli.sh`?

**Comment by tomasherman** [comment]
can you pls create a minimized offending swagger definition? that would make it easier for us to fix this ...

**Comment by BBartosz** [comment]
@tomasherman 

```
{
  "swagger": "2.0",
  "info": {
    "title": "My API",
    "description": "my test api",
    "version": "v1"
  },
  "x-basePath": "{environment}:8080/test-api/{version}",
  "schemes": [
    "https"
  ],
  "consumes": [
    "application/json"
  ],
  "produces": [
    "application/json"
  ],
  "definitions": {
    "option": {
      "description": "some description",
      "properties": {
        "number": {
          "default": 1,
          "minimum": 1,
          "maximum": 8,
          "description": "Some Number",
          "example": 1,
          "type": "integer"
        }
      }
    },
    "response": {
      "description": "some description",
      "properties": {
        "number": {
          "default": 1,
          "minimum": 1,
          "maximum": 8,
          "description": "Some Number ",
          "example": 1,
          "type": "integer"
        }
      }
    }
  },
  "/search": {
    "post": {
      "operationId": "post_search",
      "responses": {
        "200": {
          "description": "some search response",
          "examples": {},
          "schema": {
            "$ref": "#/definitions/response"
          }
        }
      },
      "parameters": [
        {
          "schema": {
            "$ref": "#/definitions/option"
          },
          "in": "body",
          "name": "body",
          "description": "search request",
          "required": true
        }
      ]
    }
  }
}
```

Im surprised noboby encouner this problem before which makes me hope there is some obvious misusage of guardrail by me. So generated case class will be `case class Option(something)` so anywhere guardrail will generate code which returns or take as parameter SCALA `Option[+A]` from package scala we will have exception that option doesnt take parameters because in scope of generated code compiler will take first our generated case class with name Option instead scala.Option[+A]. Same will happen if you will name definition with any other keyword like `list`, `exception`, or literally any keyword from scala standard library. Example:

```
"definitions": {
    "list": {
      "description": "some description",
      "properties": {
        "number": {
          "default": 1,
          "minimum": 1,
          "maximum": 8,
          "description": "Some Number",
          "example": 1,
          "type": "integer"
        }
      }
    },
```
Simplest solution which will work for my case i see now is anywhere guardrail is generating scala Option as return type or input parameter, we need to add prefix with scala package before. 

Example:

```
  object DoubleNumber     { def unapply(value: String): scala.Option[Double]     = Try(value.toDouble).toOption    }
```
Instead right now guardrail generates:

```
  object DoubleNumber     { def unapply(value: String): Option[Double]     = Try(value.toDouble).toOption    }
```
But adding `scala.` prefix to scala Option solves problem for my particular case, not for general case where you can define definition with some other reserved name in language(i.e. list). So general solution will be add possibility for user to define prefix to all names generated by guardrail. So I want to specify: add `_` before any class you generate from my swagger docs. I cant check how to do it exactly now, I will be able to check that tomorrow.  But I have other question, @tomasherman do you understand the issue now and if your PR is solving that or your PR is not related? Im talking about #434

**Comment by tomasherman** [comment]
so i just tried it and it does NOT compile there is still a lot of work to be done if we want to support this - 

i used following swagger:

```yaml
swagger: '2.0'
info:
  title: Polymorphism nested example
  version: 1.0.0
produces:
  - application/json
paths:
  /foo:
    get:
      operationId: getPet
      responses:
        200:
          description: Return the details about the pet
          schema:
            $ref: '#/definitions/Option'
definitions:
  Option:
    properties:
      value:
        type: number
```

**Comment by tomasherman** [comment]
so basically to fix this we would need to prefix all `Option` and `List` etc with `scala.` ... do we want to go that way before i sink a lot of effort into that branch @blast-hardcheese to me it does make sense

**Comment by tomasherman** [comment]
(i only intend to implement this for circe as i do not care all that much for other platforms)

**Comment by tomasherman** [comment]
i tried to fix this but didn't make it - maybe i will get back to it but if someone else wants to continue, def go fo it

**Comment by blast-hardcheese** [comment]
@tomasherman I'd need to check the OAI specification, but I wonder if something similar to `mapping` exists, otherwise `x-jvm-class-name` may be the cleanest option with the least changes required.

One annoying thing would be `$ref`s, as currently we just assume the last chunk of the pointer is the class name.

This would be additionally beneficial, on the other hand, by enabling exhaustivity testing for unused models.

@BBartosz Would this workaround work for you?

**Comment by blast-hardcheese** [comment]
(I'll also clarify, the reason I think vendor extensions are acceptable here is because guardrail is already in charge of extracting class names from specification files using some heuristics, so this would just influence the final outcome.

One additional point is that in untyped languages, there's no direct need to name these structures at all, beyond enforcing validation rules. I think that gives us many liberties to offer the best experience to the user, without complex language specific rewriting rules.)

**Comment by tomasherman** [comment]
isn't the solution much simpler than adding extensions? We simply need to qualify all names in the generated code (when generating DTOs)

**Comment by tomasherman** [comment]
anyway i would be really sad if this issue blocked #434 - just saying ;)

**Comment by blast-hardcheese** [comment]
@tomasherman I'm looking for an escape hatch to handle all the cases that we haven't seen yet 😉 

I've encountered specifications where `operationId: test` and the response entity is already called `TestResponse`, so when the code gets generated we end up with:

```scala
object TestResponse {
  case class OK(
    value: TestResponse
  ) extends TestResponse
}
```

#434 definitely resolves this, by turning it into
```scala
object TestResponse {
  case class OK(
    value: _root_.polymorphismNested.client.akkaHttp.definitions.TestResponse
  ) extends TestResponse
}
```

but I'm optimizing for the situation where if we find other situations similar to the `Option` case, a user could find the bug, open an issue, then use `x-jvm-class-name: _Option` to work around it immediately and keep rolling instead of being blocked.

_**That being said,**_ `x-jvm-class-name` is equivalent to just renaming the definition in the specification, so they could just do that.

_**That being said,**_ continually changing the names from one to the next would increase the search space for collisions for other languages that may be added in the future (for instance, "Gah, `Option` conflicts with Scala. _renames to `Optional`_ Gah, that conflicts with Java. _renames to `Maybe`_ Gah, that conflicts with Haskell)

**All that being said**, prefixing stdlib terms with their fully-qualified paths is probably a good idea. It just assumes that Library authors won't arbitrarily change the FQN for classes, like `cats.FunctionK` -> `cats.arrow.FunctionK`

---

Pardon the long response, but hopefully this clarifies my position around escape hatches and trying to keep folk moving forward instead of getting blocked until a bugfix can get released.

**Comment by tomasherman** [comment]
I def. appreciate long responses ;)

I actually thought about it a little bit harder and i think better approach is to avoid wild card imports of DTOs (or actually any imports of DTOs) and instead using fully qualified types where needed - but that's more work because i don't think guardrail is ready for that yet - i wasn't able to find an easy way to add information into appropriate classes

**Comment by blast-hardcheese** [comment]
So, circling back to see whether this can be included with 0.54.3, https://github.com/twilio/guardrail/issues/438#issuecomment-551791591's

> I think generating tool should give me some flexibility how i want generated code to look like.

is leaning me heavily towards `x-${lang}-class-name`.

Revisiting the specification [posted earlier in this thread](https://github.com/twilio/guardrail/issues/438#issuecomment-552116771), the fact that the definition is named `option` doesn't in fact generate a conflict, see:

```scala
case class option(number: Option[BigInt] = None)
```

That being said, capitalizing that definition, it does in fact generate an unfortunate issue:
```scala
case class Option(number: Option[BigInt] = None)
```

I'm thinking that `x-${lang}-class-name` makes the most sense because:

1) It's very easy to do, and will root out the last few places where we just guess what the class name should be instead of having a direct path to the authoritative class name.
2) It mirrors `x-${lang}-type` in the sense that it permits overriding types to define additional language semantics (`String @@ NonEmpty`, or domain-specific classes that serialize to primitive types that match the wire protocol)
3) Carefully constraining the definition types as we've done recently goes a long way to reducing ambiguity, but we can see here that it still doesn't help in this case (`"Option": { ... }`) without threading through `Predef`, which then locks us in to particular versions of Scala, which I've tried very hard to prevent.
  Permitting control of the generated class name instead of trying to do some heuristic matching and automatic adjusting puts control of name conflicts in the hands of the user, which matches the intent of other escape hatches (`x-server-raw-response`, `x-${lang}-type`, etc).
4) Alternate constructions of completely isolating all possible type conflicts, like a type function that takes all primitive types as arguments so we have complete control of catching all name conflicts would be extraordinarily cumbersome to use. There may be some useful research that can be done here more broadly, but for the time being I don't see the benefit of delaying until something may be found to solve this problem elegantly.

To summarize, I think we can't resolve this with package constraints alone, but that's OK. Does anyone want to take a shot at implementing `x-${lang}-name`?
